pkg/pillar/volumehandlers/csihandler.go

```diff
@@ -0,0 +1,331 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//go:build kubevirt
+
+package volumehandlers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/lf-edge/edge-containers/pkg/registry"
+	zconfig "github.com/lf-edge/eve-api/go/config"
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/kubeapi"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	kerr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/rest"
+)
+
+const imageToQcowScript = "/opt/zededa/bin/copy-image-to-qcow.sh"
+
+type volumeHandlerCSI struct {
+	commonVolumeHandler
+	useVHost bool
+	config   *rest.Config
+}
+
+// NewCSIHandler is a constructor for the kubernetes CSI handler.
+func NewCSIHandler(common commonVolumeHandler, useVHost bool) VolumeHandler {
+	return &volumeHandlerCSI{
+		commonVolumeHandler: common,
+		useVHost:            useVHost,
+	}
+}
+
+func (handler *volumeHandlerCSI) GetVolumeDetails() (uint64, uint64, string, bool, error) {
+	pvcName := handler.status.Key()
+	handler.log.Noticef("GetVolumeDetails called for PVC %s", pvcName)
+	imgInfo, err := kubeapi.GetPVCInfo(pvcName, handler.log)
+	if err != nil {
+		return 0, 0, "", false, fmt.Errorf("GetPVCInfo failed for %s: %v", pvcName, err)
+	}
+	return imgInfo.ActualSize, imgInfo.VirtualSize, imgInfo.Format,
+		imgInfo.DirtyFlag, nil
+}
+
+func (handler *volumeHandlerCSI) UsageFromStatus() uint64 {
+	// use MaxVolSize for PVC
+	handler.log.Noticef("UsageFromStatus: Use MaxVolSize for PVC %s",
+		handler.status.Key())
+	return handler.status.MaxVolSize
+}
+
+func (handler *volumeHandlerCSI) PrepareVolume() error {
+	handler.log.Noticef("PrepareVolume called for PVC %s", handler.status.Key())
+	return nil
+}
+
+func (handler *volumeHandlerCSI) HandlePrepared() (bool, error) {
+	handler.log.Noticef("HandlePrepared called for PVC %s", handler.status.Key())
+	return true, nil
+}
+
+func (handler *volumeHandlerCSI) HandleCreated() (bool, error) {
+	handler.log.Noticef("HandleCreated called for PVC %s", handler.status.Key())
+	// Though we convert container image to PVC, we need to keep the image format to tell domainmgr
+	// that we are launching a container as VM.
+	if !handler.status.IsContainer() {
+		handler.status.ContentFormat = zconfig.Format_PVC
+	}
+	updateVolumeSizes(handler.log, handler, handler.status)
+	return true, nil
+}
+
+func (handler *volumeHandlerCSI) CreateVolume() (string, error) {
+
+	createContext := context.Background()
+
+	// we call createCancel to break volume creation process
+	// removing of partially created objects must be done inside caller
+	createContext, createCancel := context.WithCancel(createContext)
+
+	done := make(chan bool, 1)
```

Comment:
```suggestion
	done := make(chan struct{}, 1)
```

Reply:
@christoph-zededa I'm just curious: why is this approach better?

Reply:
It does not imply to the reader of the code that you have the choice to send true or false and that this makes a difference.